apps/explorer/components/Entity/EntityListItem.tsx

```diff
@@ -1,27 +1,72 @@
-'use client'
+import BigNumber from 'bignumber.js'
+import { formatDistance } from 'date-fns'
+import { upperFirst } from '@technically/lodash'
 
 import {
+  Badge,
   Link,
   Text,
   Tooltip,
   ValueCopyable,
   ValueScFiat,
   ValueSf,
 } from '@siafoundation/design-system'
+import {
+  ExplorerTransaction,
+  ExplorerV2Transaction,
+} from '@siafoundation/explored-types'
+import {
+  ArrowLeft16,
+  ArrowRight16,
+  DotMark16,
+} from '@siafoundation/react-icons'
 import {
   EntityType,
   getEntityTypeLabel,
   getTxTypeLabel,
+  humanNumber,
   TxType,
 } from '@siafoundation/units'
-import { humanNumber } from '@siafoundation/units'
-import { formatDistance } from 'date-fns'
-import { upperFirst } from '@technically/lodash'
-import BigNumber from 'bignumber.js'
-import { DotMark16 } from '@siafoundation/react-icons'
-import { EntityListItemLayout } from './EntityListItemLayout'
+
+import { getTransactionSummary } from '../../lib/tx'
+
 import LoadingTimestamp from '../LoadingTimestamp'
 
+import { EntityListItemLayout } from './EntityListItemLayout'
+
+export type TxPreviewBadgeConfig = {
+  arbitraryData: boolean
+  contractFormation: boolean
+  contractResolution: boolean
+  contractRevision: boolean
+  hostAnnouncement: boolean
+  spend: boolean
+}
+
+export function generateTxPreviewBadgeConfig(
+  tx: ExplorerV2Transaction | ExplorerTransaction
+): TxPreviewBadgeConfig {
+  const summary = getTransactionSummary(tx)
+  return {
+    arbitraryData: !!tx.arbitraryData?.length,
+    contractFormation: !!tx.fileContracts?.length,
+    contractResolution:
+      ('fileContractResolutions' in tx &&
+        !!tx.fileContractResolutions?.length) ||
+      ('storageProofs' in tx && !!tx.storageProofs?.length),
+    contractRevision: !!tx.fileContractRevisions?.length,
+    hostAnnouncement: !!tx.hostAnnouncements?.length,
+    spend:
+      (!!summary.sc.length || !!summary.sf.length) &&
```

Comment:
This check could include the resolution subtypes and also distinguish between siafund and siacoin transfers like in `getV2TransactionType` .